### ISSUE TITLE
There's no function for doing an OR in WHERE with the Sql class

### ISSUE THREAD

**Issue by bladeoflight16** [CollaboratingPlatypus/PetaPoco#472]
Suppose I want this query:

```
SELECT *
FROM MY_TABLE
WHERE
    (A = @0 AND B = @1)
    OR (A = @2 AND B = @3)
    OR (A = @4 AND B = @5)
```

Now say that I don't know ahead of time how many pairs like that I'll be querying for. For a query with clauses that vary dynamically like that, using a `Sql` object is the natural choice.

But doing it is ugly:

```
var sql = new Sql("SELECT * FROM MY_TABLE");

var firstThing = myListOfThings.First();
sql.Where("(A = @0 AND B = @1)", firstThing.A, firstThing.B);

foreach (var o in myListOfThings.Skip(1))
{
    sql.Append("OR (A = @0 AND B = @1)", o.A, o.B);
}
```

A whole lot better would be a dedicated function for `OR`ing the conditions together:

```
var sql = new Sql("SELECT * FROM MY_TABLE");

foreach (var o in myListOfThings)
{
    sql.WhereOr("A = @0 AND B = @1", o.A, o.B);
}
```

**Comment by asherber** [comment]
I agree that something like this would be nice, but while I can't speak for the project owner, historically PetaPoco's SQL builder has been "really basic" (to quote the [old documentation](https://github.com/CollaboratingPlatypus/PetaPoco/wiki/Old-Documentation).

In your case, you could clean up your first example a little, like so:

```csharp
var sql = new Sql("SELECT * FROM MY_TABLE");

var firstThing = myListOfThings.First();
sq.Where("1 = 0");

foreach (var o in myListOfThings)
{
    sql.Append("OR (A = @0 AND B = @1)", o.A, o.B);
}
```

Not the prettiest, but at least it avoids having to treat the first item in the list separately.

**Comment by pleb** [comment]
An updated/new query builder would awesome. I'd vote for a more fluent syntax

**Comment by asherber** [comment]
[SqlKata](https://sqlkata.com) looks like a very full-featured query builder with a fluent API. It could be used with PetaPoco via extension methods, like so:

```csharp
public static class DatabaseQueryExtensions
{
    private static Compiler _compiler = new SqlServerCompiler();

    private static SqlResult Compile(this Query q) => _compiler.Compile(q);

    private static string GetPetaPocoSql(this SqlResult sqlResult)
    {
        // SqlResult.Sql has params like @p0, but PetaPoco wants just @0
        return Helper.ReplaceAll(sqlResult.RawSql, "?", x => "@" + x);
    }

    public static List<T> Fetch<T>(this Database db, Query query)
    {
        var compiled = query.Compile();
        return db.Fetch<T>(compiled.GetPetaPocoSql(), compiled.Bindings);
    }
}
```
Note that the `SqlServerCompiler` has optimizations for SqlServer — for example, enclosing identifiers in [ ]. They also have compilers for Firebird, MySql, and Postgres.

Here's the code to create the SQL that the OP wanted, using SqlKata

```csharp
var q = new Query("MY_TABLE");

foreach (var o in list)
{
    q.OrWhere(x => x.Where("A", o.A).Where("B", o.B));
}

var recs = db.Fetch<MyTable>(q);
```

**Comment by asherber** [comment]
Actually, a much better and more compact extension method would look like this:

```csharp
public static class DatabaseQueryExtensions
{
    private static Compiler _compiler = new SqlServerCompiler();

    public static Sql ToSql(this Query query)
    {
        var compiled = _compiler.Compile(query);
        var ppSql = Helper.ReplaceAll(compiled.RawSql, "?", x => "@" + x);

        return new Sql(ppSql, compiled.Bindings);
    }
}
```
Now you can just use `query.ToSql()` in any PetaPoco methods that take a `Sql`.

```csharp
var query = new Query("MyTable").Where("Foo", "Bar");
var recs = db.Fetch<MyPoco>(q.ToSql());
```

**Comment by bladeoflight16** [comment]
@asherber I would call having `OR` queries "very basic."

**Comment by asherber** [comment]
@bladeoflight16 I can't speak to the design decisions that went into this library, but it seems to me that the focus has been on turning data into POCOs, leaving it mostly up to the user how to construct the correct query. Personally, I have always seen the `Sql` class as a nice convenience for simple queries, not as a tool I rely on. Since the `Database` methods let you pass in your own SQL, you're free to use whatever method you like to build queries, which seems to be a feature, not a bug. And since there are already more comprehensive query builder tools out there, IMO it's not worth spending time adding that functionality to this library. But I'm sure a PR would be welcomed if someone disagrees and wants to work on this.

(BTW, it seems to me that the tricky part here is that as soon as you add `OR`, you also have to add some way to support parens, because otherwise you can't always make `OR/AND` do what you want.)

**Comment by asherber** [comment]
I know this doesn't solve @bladeoflight16 's desire to have better SQL builder support inside of PetaPoco, but I just put together a NuGet package that makes it easier to use SqlKata as a builder for PetaPoco. It includes the `ToSql()` method above, plus a few other things.

It's available on [NuGet](https://www.nuget.org/packages/PetaPoco.SqlKata/) and [GitHub](https://github.com/asherber/PetaPoco.SqlKata).

**Comment by pleb** [comment]
@asherber Awesome work! Make sure you add it to the docs and master read.me.
